worlds/no100f/Rules.py

```diff
@@ -0,0 +1,2777 @@
+import typing
+from typing import Callable, Dict, List, Tuple
+
+from BaseClasses import MultiWorld, CollectionState, Entrance
+from .Options import NO100FOptions
+from .names import ConnectionNames, ItemNames, LocationNames, RegionNames
+from worlds.generic.Rules import set_rule, add_rule, CollectionRule
+
+
+upgrade_rules = [
+    # connections
+    {
+        # Hub
+
+        # Manor
+        ConnectionNames.i020_i021: lambda player: lambda state: state.has(ItemNames.HelmetPower, player, 1),
+        ConnectionNames.i003_b004: lambda player: lambda state: state.has(ItemNames.SoapPower, player, 1) and state.has(ItemNames.HelmetPower, player, 1),
```

Comment:
```suggestion
        ConnectionNames.i003_b004: lambda player: lambda state: state.has_all((ItemNames.SoapPower, ItemNames.HelmetPower), player),
```
Probably not necessary to do this refactor, but would recommend using `state.has_all` in-place of two `state.has`.
Also `state.has` defaults the quantity to 1, so if the quantity you're putting in is 1, the quantity can be omitted.

Reply:
Yea that would definitely make my code a fair bit more readable.

Unfortunately with how large that refactor would be, and how much it depends on a case by case basis I don't think I could justify doing it for now.